app/Http/Controllers/Chat/ChatController.php

```diff
@@ -123,65 +124,88 @@ public function __construct()
+
+        // messages need presence
+        if ($includeMessages || $includePresence) {
+            $presence = $this->presence();
+        }
+
+        if ($includeMessages) {
+            $channelIds = array_pluck($presence, 'channel_id');
+
+            $messages = Message
+                ::with('sender')
+                ->whereIn('channel_id', $channelIds)
+                ->since($since)
+                ->limit($limit)
+                ->orderBy('message_id', 'DESC');
+
+            if ($params['channel_id'] !== null) {
+                $messages->where('channel_id', $params['channel_id']);
+            }
+
+            $messages = $messages->get()->reverse();
+
+            $response['messages'] = json_collection($messages, new MessageTransformer(), ['sender']);
         }
 
-        $messages = $messages->get()->reverse();
+        if ($includePresence) {
+            // to match old behaviour
+            $response['presence'] = $includeMessages && ($messages ?? collect())->isEmpty()
+                ? []
+                : $presence;
+        }
 
-        $silenceQuery = UserAccountHistory::bans()->limit(100);
-        $lastHistoryId = get_int($params['history_since'] ?? null);
+        if ($includeSilences) {
+            $silenceQuery = UserAccountHistory::bans()->limit(100);
+            $lastHistoryId = $params['history_since'];
 
-        if ($lastHistoryId === null) {
-            $previousMessage = Message::where('message_id', '<=', $since)->last();
+            if ($lastHistoryId === null) {
+                $previousMessage = Message::where('message_id', '<=', $since)->last();
 
-            if ($previousMessage === null) {
-                $silenceQuery->none();
+                if ($previousMessage === null) {
+                    $silenceQuery->none();
+                } else {
+                    $silenceQuery->where('timestamp', '>', $previousMessage->timestamp);
+                }
             } else {
-                $silenceQuery->where('timestamp', '>', $previousMessage->timestamp);
+                $silenceQuery->where('ban_id', '>', $lastHistoryId)->reorderBy('ban_id', 'DESC');
             }
-        } else {
-            $silenceQuery->where('ban_id', '>', $lastHistoryId)->reorderBy('ban_id', 'DESC');
-        }
 
-        $silences = $silenceQuery->get();
+            $silences = $silenceQuery->get();
 
-        if ($messages->isEmpty() && $silences->isEmpty()) {
-            return response([], 204);
+            $response['silences'] = json_collection($silences, new UserSilenceTransformer());
         }
 
-        return [
-            'presence' => $presence,
-            'messages' => json_collection(
-                $messages,
-                new MessageTransformer(),
-                ['sender']
-            ),
-            'silences' => json_collection($silences, 'Chat\UserSilence'),
-        ];
+        $isEmpty = array_reduce($response, fn ($acc, $val) => $acc + count($val), 0) === 0;
```

Comment:
Flip the check with `array_first`?

(also is it `[]` or `null` for empty response?)

```php
$isAny = array_first($response, fn ($val) => count($val) > 0) !== null;

return $isAny ? $response : response(null, 204);
```

Reply:
or we can use `response()->noContent()` and not think about it 🤔